core/app/models/spree/notification_dispatch/action_mailer_dispatch.rb

```diff
@@ -0,0 +1,55 @@
+# Default sender class for Spree::NotificationDispatch, sends via
+# ActionMailer.
+#
+# You can see the mapping from notification message types to
+# ActionMailer classes and methods in `.mailer_dispatch_table`, which
+# you can also customize to map to different mailers or methods.
+#
+# The particular argument list for each message type is
+# specified in NotificationDispatch.signatures
+class Spree::NotificationDispatch::ActionMailerDispatch
+  class_attribute :mailer_dispatch_table
+  # values are pairs of fully qualified mailer class names, and method names
+  self.mailer_dispatch_table = {
+    carton_shipped: ['Spree::CartonMailer', :shipped_email],
+    order_confirm: ['Spree::OrderMailer', :confirm_email],
+    order_cancel: ['Spree::OrderMailer', :cancel_email],
+    order_inventory_cancel: ['Spree::OrderMailer', :inventory_cancellation_email],
+    reimbursement_processed: ['Spree::ReimbursementMailer', :reimbursement_email]
+  }
```

Comment:
I think it important to be able to switch out an ActionMailer back-end for another back-end (likely using a third party HTTP API), _without_ having to copy-and-paste or reproduce standard logic around notifications -- such as suppressing a certain notification event. You should be able to turn off a certain notification event without worrying about whether it's an ActionMailer or another delivery mechanism, or worrying that someone later changing the back-end delivery method will break your attempt to disable certain messages. 

Additionally, I think sub-classing is really no better than class_eval, just like class_eval it exposes the entirety of internals as dependencies of your customization. It's safest if you can write a new non-ActionMailer backend _without_ having to sub-class -- composition over inheritance, this keeps the contractual API as clear and as small as possible. 

So splitting the logic between the NotificationDispatch and the particular back-end, by default ActionMailerDispatch, was an attempt to meet these design goals. 

The lambda stuff definitely can be gotten rid of.  I don't think the lambda stuff knows too much about the actual mailers -- _all_ the lambda stuff knows about is the _method signature_ of the particular notification, it doens't actually know anything at all about the actual mailers.  Prior to this PR, the _only_ documentation or logic of "what is the method signature of a particular notification event" was to look at each individual ActionMailer class.  With the idea of being able to swap out an ActionMailer back-end for another delivery mechanism, I think this is unacceptable -- there needs to be a single central place that specifies exactly what signatures your custom back-end delivery needs to respond to. That's the only point of the "lambda stuff".   It could be dispensed with  if you don't think that's neccesary -- but I think there needs to be an easy  (and test-verifiable) answer to "I want to implement a non-ActionMailer back-end, exactly what messages/signatures does it need to respond to in order to handle all notifications", and the "lambda stuff" was the simplest possible solution to that I could think of. 

Every element of this design is actually intentional to meet a particular design goal. It does look like several classes that each don't do much -- but I think this is what you get when you want simple objects with clean API's composed together (instead of inheritance) -- I think it's actually the sign of a good OO design ala Sandi Metz et al.